[Elixir] Concept exercise: conditionals

neenjaw opened the pull request:
This introduces the concept exercise of conditions, similar to C#'s enums concept exercise, but pared down a little to limit to one conditional form and the use of atoms.  I think this would be an interesting starting point to use for an _ongoing narrative_ to build on.

And by working through this, I think it elicited a need for a `booleans` exercise to arise before this one.

New files introduced:
```text
.
├── .docs
│   ├── after.md
│   ├── hints.md
│   ├── instructions.md
│   └── introduction.md
├── .formatter.exs
├── .gitignore
├── lib
│   └── log_level.ex
├── .meta
│   ├── config.json
│   ├── design.md
│   └── example.ex
├── mix.exs
├── README.md
└── test
    ├── log_level_test.exs
    └── test_helper.exs
```

Also has some administrative updates for config, concepts, etc.

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/lib/log_level.ex:
```diff
@@ -0,0 +1,3 @@
+defmodule LogLevel do
+  # TODO: implement the to_label/1 function
```
I would suggest to add a method stub here. We didn't use stubs for the basics exercise as we wanted students to learn about defining methods from scratch, but as we know students will have completed the `basics` exercise at this point, I would suggest to put in a stub here. This will allow a student to focus on the actual concept instead of the routine task of defining the method themselves. It also helps the students that will be using the online editor.

Stubs should ideally give be defined such that the student will get a nice and descriptive error message. Usually, this means that the stub should be valid, compiling code, as compiler errors can sometimes be tricky to parse for students. How this is done is different per language. See [this F# example](https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/strings/Strings.fs) for one example.

neenjaw replied:
I can do that and then have it raise a an error with a relevant message

ErikSchierboom replied:
Awesome!

ErikSchierboom commented on reference/concepts/conditionals.md:
```diff
@@ -4,3 +4,11 @@ This covers the general concept of `if`/`else`.
 Tracks should teach non-standard ways that conditionals are used in their languages (e.g. the ternary operator, postfix `if` in Ruby, `unless` in Ruby, `x if y else z` in Python).
 
+
+#### Implementations
```
Could you also add the parent `## Exercises` section and within that add an entry for your exercise (see [this example](https://github.com/exercism/v3/blob/master/reference/types/boolean.md#exercises)).

neenjaw commented on languages/elixir/exercises/concept/conditionals/.docs/instructions.md:
```diff
@@ -0,0 +1,30 @@
+
+### 1. Return the logging code label
+
+Implement the `LogLevel.to_label/1` function to return the label of a log's line from an integer code:
```
Sure, it sounds better when I re-read it

neenjaw commented on languages/elixir/exercises/concept/conditionals/.meta/example.ex:
```diff
@@ -0,0 +1,13 @@
+defmodule LogLevel do
+  def to_label(level) do
```
Hmmm, will ponder on this one while at work today. 

Is there a related second function to implement here? `Maybe a to_code/1?`

ErikSchierboom replied:
Sure, whatever makes sense (I'm no Elixir expert).

neenjaw replied:
I think I have something, see the new task 'Send an alert' to implement `send_alert?/1`

neenjaw commented:
updates based on code review: 
- config.json, removed extre prereq
- README.md, removed generated readme
- hints.md, added hint for third section
- instructions.md, typo fix, third section
- introduction.md, typo fix
- example.ex, implementation of third part
- log_level.ex, now with function stubs
- log_level_test.ex, add describe block, pending tag, third section test
- conditionals.md, add exercises header and content

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/after.md:
```diff
@@ -0,0 +1,19 @@
+You can use atoms whenever you have a set of constants you want to express. Using atoms opens a type-safe way of interacting with constant values. An atom is defined by its name, follwing the atom format:
```
```suggestion
You can use atoms whenever you have a set of constants you want to express. Using atoms opens a type-safe way of interacting with constant values. An atom is defined by its name, following the atom format:
```

ErikSchierboom replied:
The `hints.md` file has a nice link to a document on atoms. Could you also link to that document somewhere in this document?

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/introduction.md:
```diff
@@ -0,0 +1,29 @@
+
+Often, we want to write code that can branch based on a condition. While there are many ways to do this in Elixir, one of the simplest ways is using a `cond` function.
+
+At its simplest, a `cond` follows the first path that evaluates to _true_:
```
Maybe to further make the introduction concise, maybe just have the second example? I feel that would be enough.

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/introduction.md:
```diff
@@ -0,0 +1,29 @@
+
+_Atoms_ are internally represented by an integer in a lookup table, which are set automatically.  It is not possible to change this internal value.
+
+Often, we want to write code that can branch based on a condition. While there are many ways to do this in Elixir, one of the simplest ways is using a `cond` function.
```
Could you also add the `cond` description to the `after.md` document?

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/hints.md:
```diff
@@ -0,0 +1,20 @@
+### 1. Return the logging code label
+
+- You can use a [`cond/1` function][cond] to elegantly handle the various log codes.
+- You can use `===/2` to compare integers for strict type equality
```
Is there a reference somewhere that you can point to instead of having the actual code here?

ErikSchierboom replied:
Added a period at the end for consistency.

```suggestion
- You can use `===/2` to compare integers for strict type equality.
```

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/hints.md:
```diff
@@ -0,0 +1,20 @@
+### General
+
+- The [atom type is described here][atom]
```
Added a period at the end for consistency.

```suggestion
- The [atom type is described here][atom].
```

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/after.md:
````diff
@@ -0,0 +1,19 @@
+variable = :atom
+```
+
+_Atoms_ are internally represented by an integer in a lookup table, which are set automatically.  It is not possible to change this internal value.  It is generally considered to be an [anti-pattern][anti-pattern] to dynamically create atoms from user supplied input.
````
This anti-pattern remark is a perfect example of something that you would indeed want to mention in the `after.md` document and not in the `introduction.md` document. Well done!

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/instructions.md:
```diff
@@ -0,0 +1,39 @@
+
+### 3. Send an alert
+
+Implement the `LogLevel.send_alert?/1` function to determine if an alert should be send based on the log line's label.
```
I might have missed it, but is it explained somewhere which log levels should have an alert sent?

neenjaw replied:
Will clarify

neenjaw replied:
Added in revision

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/instructions.md:
```diff
@@ -0,0 +1,39 @@
+- 6 (error)
+- 7 (fatal)
+
+You have two tasks:
```
```suggestion
You have three tasks:
```

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/instructions.md:
````diff
@@ -0,0 +1,39 @@
+# => :unknown
+```
+
+### 3. Send an alert
````
I like the new addition!

ErikSchierboom commented on languages/elixir/reference/README.md:
```diff
@@ -167,4 +167,4 @@ The concept exercises use the following concepts:
 | concept | interpretation |
 | --- | --- |
-| `default-arguments` | Extract Elixir specifics from [official guide](https://elixir-lang.org/getting-started/modules-and-functions.html#default-arguments). |
+| `conditionals` | Intro to elixir `cond/1` function and atoms. |
```
As corresponding to the `config.json`, there are two Concepts introduced:

```suggestion
| `atoms` | Intro to atoms. |
| `conditionals` | Intro to elixir `cond/1` function. |
```

neenjaw replied:
Okay, so separate them, but they can still point to the one exercise. I can do that

ErikSchierboom replied:
This document is just an overview of all the concepts that are listed in the `config.json`. As such, there is no direct link to an exercise here. That is done in the `exercises/concept/README.md` file.

ErikSchierboom commented on reference/concepts/conditionals.md:
```diff
@@ -4,3 +4,18 @@ This covers the general concept of `if`/`else`.
+
+This exercise takes log message codes and return a symbol (atom) to characterize it. The reference implementation (Elixir) teaches:
+
+- conditional blocks using the `cond/1` function
```
Maybe describe this in a language-independent way?

gwaerondor commented on languages/elixir/exercises/concept/conditionals/.meta/design.md:
```diff
@@ -0,0 +1,47 @@
+
+- Pattern matching.
+- Function overloading.
+- `case` function.
```
case _statements_ I think would be better nomenclature

neenjaw replied:
I will update to match the getting started guide `control flow structure`

neenjaw replied:
I think one point I am trying to get across to student is that everything is a function, even if it doesn't look like it on the surface.

neenjaw commented:
Build is just failing on the known merge conflict, will be able to resolve one the `basics` exercise is merged.

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/after.md:
```diff
@@ -0,0 +1,31 @@
+
+[_Atoms_][atom] are internally represented by an integer in a lookup table, which are set automatically.  It is not possible to change this internal value.  It is generally considered to be an [anti-pattern][anti-pattern] to dynamically create atoms from user supplied input.
+
+[_Atoms_][atom] are also often used to represent finite states. As many functions in Elixir's standard library return an atom to annotate the result:
```
```suggestion
[_Atoms_][atom] are also often used to represent finite states. Many functions in Elixir's standard library return an atom to annotate the result:
```

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/after.md:
````diff
@@ -0,0 +1,31 @@
+# => :error
+```
+
+When we want to have branching code, we can the use of the `cond/1` function:
````
```suggestion
When we want to have branching code, we can use the `cond/1` function:
```

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/after.md:
````diff
@@ -0,0 +1,31 @@
+end
+```
+
+At least one clause should evaluate to `true` or an error will be raised.
````
Should we perhaps make it clear that the error will be run-time? I don't know if that makes sense, so feel free to ignore.

```suggestion
At least one clause should evaluate to `true` or a run-time error will be raised.
```

neenjaw replied:
the dreaded runtime vs run-time!

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/hints.md:
```diff
@@ -0,0 +1,21 @@
+### General
+
+- The [atom type is described here][atom]
```
Just for consistency :)

```suggestion
- The [atom type is described here][atom].
```

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/instructions.md:
```diff
@@ -0,0 +1,41 @@
+
+An engineer wants to be alerted when some events occur in the log.  They would like to be notified if an `:error`, `:fatal`, or `:unknown` event occurs.
+
+Implement the `LogLevel.send_alert?/1` function to determine if an alert should be send based on the log line's label.
```
```suggestion
Implement the `LogLevel.send_alert?/1` function to determine if an alert should be sent based on the log line's label.
```

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/hints.md:
```diff
@@ -0,0 +1,21 @@
+
+### 1. Return the logging code label
+
+- You can use a [`cond/1` function][cond] to elegantly handle the various log codes.
```
This line gives away the solution. Maybe you could rephrase a bit to not include the solution?

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/hints.md:
```diff
@@ -0,0 +1,21 @@
+
+### 3. Send an alert
+
+- You can use a [`cond/1` function][cond] to compare the code's label, determine whether an alert should be sent.
```
This line gives away the solution. Maybe you could rephrase a bit to not include the solution?

neenjaw replied:
I just shortened it to something more similar to the first hint.

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/.docs/introduction.md:
```diff
@@ -0,0 +1,21 @@
+Elixir's `atom` type represents a fixed constant. An atom's value is simply its own name.  This gives us a type-safe way to interact with data.  Atoms can be defined as follows:
```
This introduction is excellent! It is clear and concise. Love it!

neenjaw commented:
@ErikSchierboom - I have made the changes requested and I think resolved the merge conflicts.

@Cohen-Carlisle - yet another PR for your list when you have a chance :)

ErikSchierboom commented on languages/elixir/exercises/concept/conditionals/lib/log_level.ex:
```diff
@@ -0,0 +1,9 @@
+defmodule LogLevel do
+  def to_label(level) do
+    raise "to_label/1 not yet implemented"
```
As we want this message to be as clear as possible to the student, maybe do:

```suggestion
    raise "Please implement the to_label/1 method"
```

If you like this, the same comment applies to the other stub method.

neenjaw commented:
@Cohen-Carlisle I'm going to merge this just so it's in, we can do a revision pass later
